### DISCUSSION TITLE
Memgpt

### DISCUSSION THREAD

**Discussion by MindsightsAI** [crewAIInc/crewAI#1104] [Feature Requests]
How to use with memgpt as one of agents ?

**Comment by greysonlalonde** [comment]
@MindsightsAI see [this active langchain issue](https://github.com/langchain-ai/langchain/issues/12091)

**Comment by joaomdmoura** [comment]
Independently of LangChain, I would love to add MemGPT! Def something that has been on my radar!Adding as feature accepted

**Comment by D4zz4** [comment]
That would be awesome!

**Comment by Coriolan-Bataille** [comment]
Any news on this part?

**Comment by cheepon** [comment]
What is the current progress？

**Comment by imrohankataria** [comment]
UPDATES?

**Comment by sanikamnjoshi** [comment]
Has anyone tried using letta with crewai?

**Comment by raeros** [comment]
any news?

**Comment by kinthaiofficial** [comment]
MemGPT's paged memory model is genuinely clever — the idea of treating the context window like virtual memory with explicit page-in/page-out is a clean abstraction. The gap we kept running into: the paging decision is made by the same LLM that's paying the cost of the page, which creates a circular dependency when the model itself starts losing track of what it has forgotten.

What we ended up doing in KinthAI: separate the memory layer from the reasoning layer. A dedicated lightweight model handles consolidation ("dreaming" — deduction + induction specialists that run offline, not in-path). The primary agent just gets a structured summary + entity index, never raw memory dumps. 

The key metric we track is "context efficiency ratio" — referenced tokens / total context tokens. When it drops below 40% you're mostly paying to carry dead context. Running proactive compaction before that threshold (rather than reactive after the window fills) cut our per-task cost by about 35%.

We wrote up the full architecture here if it's useful: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

The MemGPT team is doing important work on this — curious whether you're seeing the paging overhead become a bottleneck at scale.

**Comment by kinthaiofficial** [comment]
The MemGPT concept — giving agents a structured memory hierarchy — solves a real problem with context-limited agents. The key insight is treating memory as a tiered system rather than a flat context window.

**Core memory** — always in context, high-signal facts about the task and user. Typically kept small (< 500 tokens).

**Archival memory** — vector store with semantic retrieval. Agent explicitly queries it when it needs historical context.

**Recall memory** — recent conversation history, retrieved chronologically. Different from archival because you want recency weighting, not just relevance.

The self-editing aspect (agents rewriting their own core memory) is the most interesting part architecturally — it means agents can decide what's worth remembering across sessions, which is a form of meta-learning.

Practical challenges when implementing this in CrewAI:

- **Memory consistency across agents**: if Agent A updates shared memory, Agent B needs to invalidate its cached view. Requires a pub/sub or invalidation mechanism.
- **Conflicting memories**: two agents might write contradictory facts. You need a reconciliation strategy (last-write-wins, merge, flag-for-human).
- **Memory poisoning**: if an agent writes incorrect facts to persistent memory, those errors persist and compound. Validation before write helps.

We've written about persistent memory architectures for agents and why the design choices are surprisingly similar across conversational AI and agent systems: https://blog.kinthai.ai/why-character-ai-forgets-you-persistent-memory-architecture

**Comment by jingchang0623-crypto** [comment]
MemGPT integration with CrewAI is an interesting idea — combining structured agent roles with persistent memory.

## What We Learned About Agent Memory

We have been running 6 AI agents in production for a few months, and memory management turned out to be the hardest problem. Not coordination, not tool use — **memory**.

### The Three Memory Layers We Use

1. **Session Memory** — Short-term context within a single task execution. This is what CrewAI already handles well.

2. **Episodic Memory** — Records of past task executions. We store these as markdown files organized by date:
```python
# After each crew execution
with open(f"memory/{today}.md", "a") as f:
    f.write(f"## {task_name}\n")
    f.write(f"Result: {result}\n")
    f.write(f"Tokens: {token_count}\n")
    f.write(f"Duration: {elapsed}\n")
```

3. **Semantic Memory** — Long-term knowledge that persists across sessions. This is where MemGPT-style approaches shine.

### The Key Insight

The biggest win was not the memory technology itself, but **knowing when to retrieve what**. Our agents now follow this pattern:

```
Task arrives → Check episodic memory for similar past tasks
             → If found, use that as starting context
             → Execute task with session memory
             → Store result in episodic memory
             → Extract learnings → semantic memory
```

### MemGPT + CrewAI Suggestion

If you want to integrate them, consider this architecture:
- **MemGPT** handles the memory tier management (what stays, what goes, what gets summarized)
- **CrewAI** handles the agent orchestration and task routing
- Connect them through a shared memory tool that any CrewAI agent can call

We benchmarked several memory approaches and wrote about it at [miaoquai.com](https://miaoquai.com). The TL;DR: simple file-based memory beats vector DB for most use cases under 10k documents.

**Comment by citw2** [comment]
You're looking to integrate MemGPT as one of your agents. To maintain context and avoid re-feeding history, consider using a persistent memory layer. Full disclosure: SAIHM is the project I'm posting on behalf of. SAIHM allows you to store and recall context by keyword, so your agent only loads relevant information, reducing token usage. You hold the encryption keys, ensuring data sovereignty.

More on how that works: https://t.saihm.coti.global/r/gh-alt
